### PR TITLE
[HGCal] RecHits calibration using regional e/m factors

### PR DESCRIPTION
**Pull request by apsallid**
#### PR description:

The calibration of HGCal reconstructed hits is a well documented procedure and details can be found in this [talk](https://indico.cern.ch/event/740826/contributions/3058501/attachments/1678687/2696151/CalibratedRecHits.pdf) and in this [documentation](http://hgcal.web.cern.ch/hgcal/HitCalibration/hitCalibration.html).

After detailed studies, which can be found in [1], we extend the thickness correction factors to 
cover also CE-H silicon and scintillator regions. 

#### PR validation:

We have tested the PR using the usual HGCal validation package as well as the normal HGCal 
reco tools. Results are gathered in these slides [1].

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not a backport.

[1] http://apsallid.web.cern.ch/apsallid/Regionalemfactors.pdf

@rovere @felicepantaleo @cseez @pfs @amartelli

### CODE REVIEW

**Comment by kpedro88** [comment]
ping @cmsbuild

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
There are changes in the HGCal RecHit distribution, which is probably expected:
HGCEE rechits
HGCHEF rechits
HGCHEB rechits

Please confirm if these changes go in the desired directions.

**Comment by mariadalfonso** [comment]
These changes in the DeDx will nee to be propagated to the HGC-L1 @jbsauvan @cerminar 
and are relevant also for the HFNose (L1 + Offline)

**Comment by cseez** [comment]
Bit of a misunderstanding here. There have been no changes to the dEdx 
weights. The changes to the RecHit calibration is, in effect, a 
correction to how we handle the sensor signals. The RecHit energy is a 
product of the dEdx weights and the number of MIPs passing through the 
sensors. We are correcting for small discrepancies due to an imperfect 
estimation of that number.

**Comment by rovere** [comment]
> There are changes in the HGCal RecHit distribution, which is probably expected:
> HGCEE rechits
> HGCHEF rechits
> HGCHEB rechits
> 
> Please confirm if these changes go in the desired directions.

Ciao @kpedro88 thanks for checking the plots.
Unfortunately, I do not think they are the best possible to judge the correctness of this PR.
I'd kindly ask @apsallid to produce the _usual calibration plots_ for HGCAL in the different regions, with and without this PR, to show how the energy of the rechits in different regions behaves.

**Comment by kpedro88** [comment]
Yes, having calibration plots with higher statistics is the best option. However, I will still look at the PR comparison plots, since only those plots are guaranteed to be made with the exact code introduced in the PR (there is always some rare case where a file was not committed properly, etc.).

**Comment by apsallid** [comment]
Hi @kpedro88 , @rovere 

In the pdf file in the description I have produced the usual calibration plots for all regions as usual with and without PR, using not only our reco-prod tools but also the validation tools from the DQM.

**Comment by kpedro88** [comment]
@apsallid thanks, the PR plots look consistent with slides 9 and 10 to me (slight shift up in EE, slight shift down in HE).

**Comment by kpedro88** [comment]
+upgrade

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.h [hunk 1]
```diff
@@ -56,6 +56,8 @@ class HGCalRecHitWorkerSimple : public HGCalRecHitWorkerBaseClass {
   uint32_t rangeMask_;
 
   std::vector<double> rcorr_, rcorrNose_;
+  double rcorrscint_;
+  int deltasi_index_regemfac;
```
Comment (perrotta) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.h]: Trailing underscore, as for other class members

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- It extend the thickness correction factors in the RecHit calibration to cover also CE-H silicon and scintillator regions, as described
- Changes in the outputs of the Phase2 workflows reflect that change

**Comment by kpedro88** [comment]
+upgrade

**Comment by silviodonato** [comment]
+1

**Pull request merged**